core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +99,48 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null) {
+      return null;
+    }
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+
+    // Determine mapping between project input and output fields. If sort
+    // relies on non-trivial expressions, we can't push.
+    for (RelFieldCollation fc : collation.getFieldCollations()) {
+      if (map.getTargetOpt(fc.getFieldIndex()) < 0) {
+        return null;
```

Comment:
Do we have a test case for this?
Say we have a project a, b+1 as b, c.
The parent request trait collation on a, b, c.
In this case, we can't pass it down.

Reply:
Good point. Add one test case for trait push down `testSortProjectOnRexCall` and one for trait derivation `testSortProjectDeriveOnRexCall`